### PR TITLE
Create an eval-only script for existing ckpts

### PR DESCRIPTION
**Pull request by liujch1998**
This PR adds `scripts/eval.py`, which evaluates one or more existing ckpts while bypassing the training steps.

It seems impossible to backfill evals back to the original wandb run, because "step" must always increase. Therefore, this script logs things to a new wandb run.

To run this script, you can specify a single ckpt in as `--load_path`, or all ckpts under a directory if you remove the `/stepXXX` part.

### CODE REVIEW

File: configs/peteish1-weka.yaml [hunk 1]
```diff
@@ -108,35 +108,35 @@ eval_interval: 1000
...
+        dolma_wiki-validation:
+          - /weka/oe-training-default/ai2-llm/eval-data/perplexity/v3_small_dolma2-tokenizer/dolma_wiki/val/part-0-00000.npy
+        ice-validation:
+          - /weka/oe-training-default/ai2-llm/eval-data/perplexity/v3_small_dolma2-tokenizer/ice/val/part-0-00000.npy
+        m2d2_s2orc-validation:
+          - /weka/oe-training-default/ai2-llm/eval-data/perplexity/v3_small_dolma2-tokenizer/m2d2_s2orc/val/part-0-00000.npy
+        pile-validation:
+          - /weka/oe-training-default/ai2-llm/eval-data/perplexity/v3_small_dolma2-tokenizer/pile/val/part-0-00000.npy
+        wikitext_103-validation:
+          - /weka/oe-training-default/ai2-llm/eval-data/perplexity/v3_small_dolma2-tokenizer/wikitext_103/val/part-0-00000.npy
```
Comment (dirkgr) [inline configs/peteish1-weka.yaml:139]: Did you mean to commit this change?

File: configs/peteish1-weka.yaml [hunk 2]
```diff
@@ -231,6 +231,228 @@ evaluators:
...
+    type: downstream
+
+  - label: arc_challenge_mc_5shot
+    type: downstream
+
+  - label: arc_challenge_mc_5shot_bpb
+    type: downstream
+
+  - label: copa_rc_0shot
+    type: downstream
```
Comment (dirkgr) [inline configs/peteish1-weka.yaml:379]: Do we care about any of the 0-shots?

File: olmo/train.py [hunk 3]
```diff
@@ -1368,8 +1368,8 @@ def close(self, exit_code: int = 0) -> None:
             gc.enable()
         else:
             gc.disable()
-        if wandb.run is not None:
-            wandb.finish(exit_code=exit_code, quiet=True)
+        # if wandb.run is not None:
+        #     wandb.finish(exit_code=exit_code, quiet=True)
```
Comment (dirkgr) [inline olmo/train.py:1372]: Debug code?

File: scripts/eval.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        )
+
+    barrier()
+
+    # Set seed.
+    seed_all(cfg.seed)
+
+    # # Construct data loader.
+    # train_loader = build_train_dataloader(cfg)
+    train_loader = None
```
Comment (dirkgr) [inline scripts/eval.py:120]: Is this always going to be `None`? If so, we don't need it.

File: scripts/eval.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    # Construct evaluators.
+    evaluators = build_evaluators(cfg, device)
+    barrier()
+
+    if cfg.load_path is None:
+        raise OLMoConfigurationError("To run eval you must provide a load_path")
+    if 'step' in cfg.load_path.split('/')[-1]:
+        load_paths = [cfg.load_path]
+    else:
+        # This globbing does not work with remote paths.
```
Comment (dirkgr) [inline scripts/eval.py:131]: How is that problem handled then?

File: scripts/eval.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        step = int(load_path.split('/')[-1].split('step')[-1])
+
+        # Initialize the model.
+        log.info("Building model...")
+        olmo_model = OLMo(cfg.model)
+        log.info(f"Total number of parameters: {olmo_model.num_params():,d}")
+        log.info(f"Number of non-embedding parameters: {olmo_model.num_params(include_embedding=False):,d}")
+        log.info(f"Peak GPU Memory (MB) before {cfg.distributed_strategy}: {int(peak_gpu_memory() or 0)}")
+
+        olmo_model.set_activation_checkpointing(cfg.activation_checkpointing)
```
Comment (dirkgr) [inline scripts/eval.py:144]: If we only ever eval, we don't need this.

File: scripts/eval.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        if param_init_fn is not None or cfg.distributed_strategy == DistributedStrategy.ddp:
+            olmo_model.reset_parameters()
+
+        log.info(f"Peak GPU Memory (MB) after {cfg.distributed_strategy}: {int(peak_gpu_memory() or 0)}")
+        log.info("Model:")
+        log.info(dist_model)
+
+        # Construct optimizer and learning rate scheduler.
+        optim = build_optimizer(cfg, dist_model)
+        scheduler = build_scheduler(cfg)
```
Comment (dirkgr) [inline scripts/eval.py:226]: We don't need optimizers and schedulers if we're just evaluating.

Comment (dirkgr) [reply scripts/eval.py:226]: So you're creating these only so that you can produce a `Trainer` object?

How hard is it to pull the stuff you need out of the `Trainer` object, so we don't have to do so many things we don't need? It makes me particularly uncomfortable that you're creating a trainer with a `None` data loader, which isn't supposed to work. It just happens to work.

**Review by dirkgr — CHANGES_REQUESTED** [review]

File: configs/peteish1-weka.yaml (hunk 1 above)
Comment (liujch1998) [reply configs/peteish1-weka.yaml:139]: Yes. I'm under the impression that we should add PPL eval back to peteish training.

File: configs/peteish1-weka.yaml (hunk 2 above)
Comment (liujch1998) [reply configs/peteish1-weka.yaml:379]: Better log them than regret & backfill :)

File: scripts/eval.py (hunk 5 above)
Comment (liujch1998) [reply scripts/eval.py:131]: Not handled. I will assume the checkpoints are on WEKA.

File: scripts/eval.py (hunk 7 above)
Comment (liujch1998) [reply scripts/eval.py:226]: The Trainer class has too many precious helper functions and it's kinda dumb to unroll them all. I do wanna keep at least a dummy Trainer object. Let me see if I can create it w/o the optim/scheduler/etc stuff.

Comment (dirkgr) [reply scripts/eval.py:222]: I think you might find that you don't need most of that stuff when you're doing inference only.

File: scripts/eval.py (hunk 5 above)
Comment (dirkgr) [reply scripts/eval.py:127]: At least throw an exception then.

**Comment by dirkgr** [comment]
Let me know when this is ready for another review?

File: scripts/eval.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+    seed_all(cfg.seed)
+
+    # Construct evaluators.
+    evaluators = build_evaluators(cfg, device)
+    barrier()
+
+    if cfg.load_path is None:
+        raise OLMoConfigurationError("To run eval you must provide a load_path")
+    elif "://" in cfg.load_path:
+        raise OLMoConfigurationError("Eval does not support remote paths. Please specify a local path or WEKA mounted path.")
```
Comment (liujch1998) [inline scripts/eval.py:125]: throwing exception for remote paths here
